### PR TITLE
realtek: Fix MMD-register access on RTL838x

### PR DESCRIPTION
**Pull request by oliv3r**
This fixes 2 bugs in MMD access on the RTL838x:
	- We need to preserve the state of the park-page because the normal PHY r/w functions expect a correct setting in RTL838X_SMI_ACCESS_PHY_CTRL_1
	- commit b53202a8c3f728c348c5376e5b5fb36af7c37744 introduced calling the hw MMD access functions with mdiobus_c45_devad(regnum) so now we need to invert this op via mdiobus_c45_regad(reg)

This was tested with MMD access register R/W for EEE functionality on the Zyxel GS1900-10HP.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

### CODE REVIEW

**Comment by oliv3r** [comment]
@svanheule a patch in a series of 'small fixes' as individual MR's.

File: target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c [hunk 1]
```diff
@@ -1891,27 +1891,27 @@ int rtl838x_write_phy(u32 port, u32 page, u32 reg, u32 val)
...
 
 	v = addr << 16 | reg;
 	sw_w32(v, RTL838X_SMI_ACCESS_PHY_CTRL_3);
 
-	/* mmd-access | read | cmd-start */
-	v = 1 << 1 | 0 << 2 | 1;
-	sw_w32(v, RTL838X_SMI_ACCESS_PHY_CTRL_1);
+	park_page = sw_r32(RTL838X_SMI_ACCESS_PHY_CTRL_1) & ((0x1f << 15) | 0x2);
+	/* mmd-access | read: BIT(2) not set | cmd-start */
+	v = BIT(1) | BIT(0);
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c:1907]: And TYPE_MMD is again set here. That's a bit redundant.

File: target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c [hunk 2]
```diff
@@ -1891,27 +1891,27 @@ int rtl838x_write_phy(u32 port, u32 page, u32 reg, u32 val)
...
+	pr_debug("%s port %d, addr %d, reg %d\n", __func__, port, addr, reg);
+	sw_w32(port << 16, RTL838X_SMI_ACCESS_PHY_CTRL_2);
 
 	v = addr << 16 | reg;
 	sw_w32(v, RTL838X_SMI_ACCESS_PHY_CTRL_3);
 
-	/* mmd-access | read | cmd-start */
-	v = 1 << 1 | 0 << 2 | 1;
-	sw_w32(v, RTL838X_SMI_ACCESS_PHY_CTRL_1);
+	park_page = sw_r32(RTL838X_SMI_ACCESS_PHY_CTRL_1) & ((0x1f << 15) | 0x2);
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c:1905]: This extracts the PARK_PAGE field _and_ hard sets the TYPE field (to MMD). AFAICT PARK_PAGE isn't actually set anywhere at all, so does this just depend on the last value it was set to before this driver runs?

Do you know what value is stored here? According to the SDK, a value of 0x1f means something along the lines of "ignore parked page". I have no idea what a "parked page" is however, but even the SDK seems to almost always use 0x1f. I've found a few occurrences of `PARK_PAGE = 0`

File: target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c [hunk 3]
```diff
@@ -1891,27 +1891,27 @@ int rtl838x_write_phy(u32 port, u32 page, u32 reg, u32 val)
...
+	/* mmd-access | read: BIT(2) not set | cmd-start */
+	v = BIT(1) | BIT(0);
+	sw_w32(v | park_page, RTL838X_SMI_ACCESS_PHY_CTRL_1);
 
 	if (rtl838x_smi_wait_op(100000))
 		goto timeout;
 
+	sw_w32(park_page, RTL838X_SMI_ACCESS_PHY_CTRL_1);
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c:1913]: Does it really matter if you leave the last only PARK_PAGE (and operation type) set in this register? Seems a bit fragile to depend on keeping state in this type of register.

**Review by svanheule — COMMENTED** [review]
I feel like this warrants two separate commits (Fixes: tags etc)

File: target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c [hunk 4]
```diff
@@ -1891,27 +1891,27 @@ int rtl838x_write_phy(u32 port, u32 page, u32 reg, u32 val)
...
 {
 	u32 v;
+	u32 park_page;
 
 	mutex_lock(&smi_lock);
 
-	if (rtl838x_smi_wait_op(100000))
-		goto timeout;
-
-	sw_w32(1 << port, RTL838X_SMI_ACCESS_PHY_CTRL_0);
```
[the remark this answers is not in the document]
Comment (oliv3r) [reply target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c:1900]: You do realize I can't answer you this question :D

File: target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c (hunk 1 above)
Comment (oliv3r) [reply target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c:1907]: That I can clean up :)

File: target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c (hunk 3 above)
Comment (oliv3r) [reply target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c:1913]: I whole-heatedly agree with you here. And I admit you've looked into this patch in far more detail then I have!

But you are right, there's a lot of assumptions happening in this entire driver ...

File: target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c (hunk 2 above)
Comment (oliv3r) [reply target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c]: the PARK_PAGE MR is now #11090 and we can move the discussion there.

**Comment by oliv3r** [comment]
So in 'v2', I kept things more conservative, and used 'logic' (which we know realtek is not), so I really need some help in figuring out (and testing) if this is all right. I'm not sure how this was tested before either.

**Pull request closed without merging**